core/src/test/scala/unit/kafka/utils/TestUtils.scala

```diff
@@ -1396,11 +1396,13 @@ object TestUtils extends Logging {
   // Note: Call this method in the test itself, rather than the @AfterEach method.
   // Because of the assert, if assertNoNonDaemonThreads fails, nothing after would be executed.
   def assertNoNonDaemonThreads(threadNamePrefix: String): Unit = {
-    val nonDaemonThreads = Thread.getAllStackTraces.keySet.asScala.filter { t =>
-      !t.isDaemon && t.isAlive && t.getName.startsWith(threadNamePrefix)
-    }
-    val threadCount = nonDaemonThreads.size
-    assertEquals(0, threadCount, s"Found unexpected $threadCount NonDaemon threads=${nonDaemonThreads.map(t => t.getName).mkString(", ")}")
+    var nonDemonThreads: mutable.Set[Thread] = mutable.Set.empty[Thread]
+    waitUntilTrue(() => {
+      nonDemonThreads = Thread.getAllStackTraces.keySet.asScala.filter { t =>
+        !t.isDaemon && t.isAlive && t.getName.startsWith(threadNamePrefix)
+      }
+      0 == nonDemonThreads.size
+    }, s"Found unexpected ${nonDemonThreads.size} NonDaemon threads=${nonDemonThreads.map(t => t.getName).mkString(", ")}", 1000)
```

Comment:
cc @divijvaidya , I found sometimes the [CI](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-15133/9/testReport/junit/kafka.server/ReplicaManagerTest/Build___JDK_11_and_Scala_2_13___testSuccessfulBuildRemoteLogAuxStateMetrics__/) is too sensitive to the non demean threads check. There are some shutdown are in async way. So you can check the failed results [here](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-15133/9/): Basically, if there are some resource not closed, all the following tests should also fail (I verified in my local env). But in the CI results, it only fail 2 of replicaManagertest, and only in jdk11. So I'm going to verify it using `waitUntilTrue` to give it some chance to wait for the threads shutdown.

Reply:
This will increase the run-time for test suite when it is leaking threads. Which should be ok, as the happy case (when we are not leaking remains the same).

Reply:
BTW, which threads do you think are closed async during Kafka server shutdown? AFAIK, we have checks in place to ensure that we wait for proper close during shutdown

Reply:
> BTW, which threads do you think are closed async during Kafka server shutdown? AFAIK, we have checks in place to ensure that we wait for proper close during shutdown

Because in the failed test cases, They are `ReplicaFetcherThread` leaked. But had a quick look, yes, we did do `awaitShutdown()` for fetcher threads. Hmm... Let me dig into further tomorrow.